kernel/aster-nix/src/vm/vmo/mod.rs

```diff
@@ -127,219 +135,370 @@ bitflags! {
-        }
-
-        let inherited_frame = inherited_frames.get(page_idx).unwrap().clone();
-
-        if self.should_share_frame_with_parent(write_page) {
-            return Ok(inherited_frame);
+    /// Decommit the page corresponding to the target offset in the Vmo.
+    fn decommit_page(&mut self, offset: usize) -> Result<()> {
+        let page_idx = offset / PAGE_SIZE + self.page_idx_offset;
+        let mut pages = self.pages.lock();
+        let is_cow_vmo = pages.is_marked(VmoMark::CowVmo);
+        let mut cursor = pages.cursor_mut(page_idx as u64);
+        if cursor.remove().is_some()
+            && let Some(pager) = &self.pager
+            && !is_cow_vmo
+        {
+            pager.decommit_page(page_idx)?;
         }
-
-        let frame = VmAllocOptions::new(1).alloc_single()?;
-        frame.copy_from_frame(&inherited_frame);
-        Ok(frame)
-    }
-
-    fn is_cow_child(&self) -> bool {
-        self.is_cow
-    }
-
-    fn should_share_frame_with_parent(&self, write_page: bool) -> bool {
-        !self.is_cow || !write_page
-    }
-}
-
-impl Vmo_ {
-    pub fn commit_page(&self, offset: usize) -> Result<()> {
-        self.inner.lock().commit_page(offset)
-    }
-
-    pub fn decommit_page(&self, offset: usize) -> Result<()> {
-        self.inner.lock().decommit_page(offset)
+        Ok(())
     }
 
-    pub fn commit(&self, range: Range<usize>) -> Result<()> {
-        let page_idx_range = get_page_idx_range(&range);
-        for page_idx in page_idx_range {
-            let offset = page_idx * PAGE_SIZE;
-            self.commit_page(offset)?;
+    /// Commit a range of pages in the Vmo, , returns the pages in this range.
+    pub fn commit(&self, range: Range<usize>, will_write: bool) -> Result<VmFrameVec> {
+        let mut pages = self.pages.lock();
+        if range.end > self.size() {
+            return_errno_with_message!(Errno::EINVAL, "operated range exceeds the vmo size");
         }
 
-        Ok(())
-    }
+        let raw_page_idx_range = get_page_idx_range(&range);
+        let page_idx_range = (raw_page_idx_range.start + self.page_idx_offset)
+            ..(raw_page_idx_range.end + self.page_idx_offset);
+        let mut frames = VmFrameVec::new_with_capacity(page_idx_range.len());
 
-    pub fn decommit(&self, range: Range<usize>) -> Result<()> {
-        let page_idx_range = get_page_idx_range(&range);
+        let is_cow_vmo = pages.is_marked(VmoMark::CowVmo);
+        let mut cursor = pages.cursor_mut(page_idx_range.start as u64);
         for page_idx in page_idx_range {
-            let offset = page_idx * PAGE_SIZE;
-            self.decommit_page(offset)?;
+            let committed_page = {
+                let is_exclusive = cursor.is_marked(VmoMark::ExclusivePage);
+                // The page has been committed, load the page directly or load it after triggering COW.
+                if let Some(committed_page) = cursor.load() {
```

Comment:
https://github.com/asterinas/asterinas/pull/632#discussion_r1525880335
> the abstracted function needs to take `CursorMut` as a parameter to ensure the advantage of next, hence its semantics will be quite strange

Indeed I don't feel it's strange. Taking a `CursorMut` is a way to indicate which page we are committing, which is similar to the page index.